File: test/github.test.ts
/**
 * Unit tests for release annotation: build-status cross-ref + image availability,
 * against captured live `gh` fixtures (mailon).
 */

import { test, expect, describe } from "bun:test"
import { join } from "node:path"
import { annotateBuild, annotateReleases } from "../src/github.ts"
import type { RawRelease, RawRun } from "../src/github.ts"

const FIX = join(import.meta.dir, "fixtures")
async function fixture<T>(name: string): Promise<T> {
  return (await Bun.file(join(FIX, name)).json()) as T
}

describe("annotateBuild", () => {
  const runs: RawRun[] = [
    {
      databaseId: 1,
      event: "release",
      headBranch: "v1.0.0",
      status: "completed",
      conclusion: "success",
      workflowName: "Build & push container image",
    },
    {
      databaseId: 2,
      event: "release",
      headBranch: "v1.1.0",
      status: "completed",
      conclusion: "failure",
      workflowName: "Build & push container image",
    },
    {
      databaseId: 3,
      event: "release",
      headBranch: "v1.2.0",
      status: "in_progress",
      conclusion: null,
      workflowName: "Build & push container image",
    },
    // A non-release push run that happens to mention a tag in its title — must
    // NOT be matched as the release build.
    {
      databaseId: 4,
      event: "push",
      headBranch: "master",
      displayTitle: "chore(release): 2.0.0",
      status: "completed",
      conclusion: "success",
      workflowName: "CI",
    },
  ]

  test("success → ready", () => {
    expect(annotateBuild("v1.0.0", runs)).toEqual({ build: "ready", runId: 1 })
  })
  test("failure → failed", () => {
    expect(annotateBuild("v1.1.0", runs)).toEqual({ build: "failed", runId: 2 })
  })
  test("in_progress → building", () => {
    expect(annotateBuild("v1.2.0", runs)).toEqual({ build: "building", runId: 3 })
  })
  test("no matching release run → none", () => {
    expect(annotateBuild("v9.9.9", runs)).toEqual({ build: "none", runId: null })
  })
  test("does not match non-release events", () => {
    // "2.0.0" only appears in a push run's title → none.
    expect(annotateBuild("2.0.0", runs)).toEqual({ build: "none", runId: null })
  })
  test("matches via displayTitle when headBranch differs", () => {
    const r: RawRun[] = [
      {
        databaseId: 9,
        event: "release",
        headBranch: "refs/tags/v3.0.0",
        displayTitle: "v3.0.0",
        status: "completed",
        conclusion: "success",
      },
    ]
    expect(annotateBuild("v3.0.0", r)).toEqual({ build: "ready", runId: 9 })
  })

  test("picks the MOST RECENT matching run (highest id), regardless of array order", () => {
    // A failed build (older id) re-run to success (newer id) for the same tag.
    // `gh run list` order is not guaranteed, so feed them oldest-last too.
    const failedThenSucceeded: RawRun[] = [
      { databaseId: 50, event: "release", headBranch: "v4.0.0", status: "completed", conclusion: "success" },
      { databaseId: 40, event: "release", headBranch: "v4.0.0", status: "completed", conclusion: "failure" },
    ]
    expect(annotateBuild("v4.0.0", failedThenSucceeded)).toEqual({ build: "ready", runId: 50 })

    // Same data, reversed array order → still the newest (id 50) wins.
    expect(annotateBuild("v4.0.0", [...failedThenSucceeded].reverse())).toEqual({
      build: "ready",
      runId: 50,
    })
  })
})

describe("annotateReleases", () => {
  test("flags prerelease, latest; carries availability when provided", () => {
    const releases: RawRelease[] = [
      { tagName: "v2.0.0", isLatest: true, isPrerelease: false },
      { tagName: "v2.1.0-rc.1", isPrerelease: true },
      { tagName: "draft-x", isDraft: true }, // dropped
    ]
    const runs: RawRun[] = [
      {
        databaseId: 1,
        event: "release",
        headBranch: "v2.0.0",
        status: "completed",
        conclusion: "success",
      },
    ]
    const availability = new Map<string, boolean | null>([
      ["v2.0.0", true],
      ["v2.1.0-rc.1", null],
    ])
    const out = annotateReleases(releases, runs, availability)

    expect(out.length).toBe(2) // draft filtered out
    const stable = out.find((r) => r.tag === "v2.0.0")!
    expect(stable.latest).toBe(true)
    expect(stable.prerelease).toBe(false)
    expect(stable.build).toBe("ready")
    expect(stable.imageAvailable).toBe(true)

    const rc = out.find((r) => r.tag === "v2.1.0-rc.1")!
    expect(rc.prerelease).toBe(true)
    expect(rc.build).toBe("none")
    expect(rc.imageAvailable).toBeNull() // unknown
  })

  test("defaults imageAvailable to null when no availability map", () => {
    const out = annotateReleases([{ tagName: "v1.0.0" }], [])
    expect(out[0]!.imageAvailable).toBeNull()
    expect(out[0]!.name).toBe("v1.0.0") // falls back to tag
  })
})

describe("annotateReleases against live mailon fixtures", () => {
  test("matches real release builds (v0.6.9 failed, v0.6.5 ready, v0.6.10 none)", async () => {
    const releases = await fixture<RawRelease[]>("mailon-releases.json")
    const runs = await fixture<RawRun[]>("mailon-runs.json")
    const out = annotateReleases(releases, runs)
    const byTag = new Map(out.map((r) => [r.tag, r]))

    // From the captured run fixture:
    //   v0.6.9 release build → failure
    //   v0.6.5 release build → success
    //   v0.6.10 has no release-event run yet → none
    expect(byTag.get("v0.6.9")!.build).toBe("failed")
    expect(byTag.get("v0.6.5")!.build).toBe("ready")
    expect(byTag.get("v0.6.10")!.build).toBe("none")
    // v0.6.10 is the latest published release.
    expect(byTag.get("v0.6.10")!.latest).toBe(true)
    // No availability supplied → all unknown.
    expect(out.every((r) => r.imageAvailable === null)).toBe(true)
  })
})

File: internal/github/github_test.go
package github

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// Unit tests for release annotation: build-status cross-ref + image
// availability, against captured live `gh` fixtures (mailon).
// Ported from tools/kc-bun/test/github.test.ts.

func loadFixture[T any](t *testing.T, name string) T {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("testdata", name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var out T
	if err := json.Unmarshal(data, &out); err != nil {
		t.Fatalf("decode fixture %s: %v", name, err)
	}
	return out
}

func annotateBuildRuns() []RawRun {
	return []RawRun{
		{DatabaseID: 1, Event: "release", HeadBranch: "v1.0.0", Status: "completed", Conclusion: "success", WorkflowName: "Build & push container image"},
		{DatabaseID: 2, Event: "release", HeadBranch: "v1.1.0", Status: "completed", Conclusion: "failure", WorkflowName: "Build & push container image"},
		{DatabaseID: 3, Event: "release", HeadBranch: "v1.2.0", Status: "in_progress", Conclusion: "", WorkflowName: "Build & push container image"},
		// A non-release push run that mentions a tag in its title — must NOT match.
		{DatabaseID: 4, Event: "push", HeadBranch: "master", DisplayTitle: "chore(release): 2.0.0", Status: "completed", Conclusion: "success", WorkflowName: "CI"},
	}
}

func TestAnnotateBuild(t *testing.T) {
	runs := annotateBuildRuns()
	check := func(tag string, wantBuild BuildStatus, wantID int64) {
		t.Helper()
		b, id := AnnotateBuild(tag, runs)
		if b != wantBuild || id != wantID {
			t.Errorf("AnnotateBuild(%q) = (%s, %d), want (%s, %d)", tag, b, id, wantBuild, wantID)
		}
	}
	check("v1.0.0", BuildReady, 1)
	check("v1.1.0", BuildFailed, 2)
	check("v1.2.0", BuildBuilding, 3)
	check("v9.9.9", BuildNone, 0)
	// "2.0.0" only appears in a push run's title → none.
	check("2.0.0", BuildNone, 0)
}

func TestAnnotateBuild_MatchesViaDisplayTitle(t *testing.T) {
	runs := []RawRun{
		{DatabaseID: 9, Event: "release", HeadBranch: "refs/tags/v3.0.0", DisplayTitle: "v3.0.0", Status: "completed", Conclusion: "success"},
	}
	if b, id := AnnotateBuild("v3.0.0", runs); b != BuildReady || id != 9 {
		t.Errorf("got (%s, %d), want (ready, 9)", b, id)
	}
}

func TestAnnotateBuild_PicksMostRecentRun(t *testing.T) {
	// A failed build (older id) re-run to success (newer id) for the same tag.
	runs := []RawRun{
		{DatabaseID: 50, Event: "release", HeadBranch: "v4.0.0", Status: "completed", Conclusion: "success"},
		{DatabaseID: 40, Event: "release", HeadBranch: "v4.0.0", Status: "completed", Conclusion: "failure"},
	}
	if b, id := AnnotateBuild("v4.0.0", runs); b != BuildReady || id != 50 {
		t.Errorf("got (%s, %d), want (ready, 50)", b, id)
	}
	// Reversed order → still the newest (id 50) wins.
	rev := []RawRun{runs[1], runs[0]}
	if b, id := AnnotateBuild("v4.0.0", rev); b != BuildReady || id != 50 {
		t.Errorf("reversed: got (%s, %d), want (ready, 50)", b, id)
	}
}

func TestAnnotateReleases_FlagsAndAvailability(t *testing.T) {
	releases := []RawRelease{
		{TagName: "v2.0.0", IsLatest: true, IsPrerelease: false},
		{TagName: "v2.1.0-rc.1", IsPrerelease: true},
		{TagName: "draft-x", IsDraft: true}, // dropped
	}
	runs := []RawRun{
		{DatabaseID: 1, Event: "release", HeadBranch: "v2.0.0", Status: "completed", Conclusion: "success"},
	}
	availability := map[string]Availability{
		"v2.0.0":      AvailPresent,
		"v2.1.0-rc.1": AvailUnknown,
	}
	out := AnnotateReleases(releases, runs, availability)
	if len(out) != 2 {
		t.Fatalf("got %d annotations, want 2 (draft filtered)", len(out))
	}
	byTag := map[string]ReleaseAnnotation{}
	for _, r := range out {
		byTag[r.Tag] = r
	}
	stable := byTag["v2.0.0"]
	if !stable.Latest || stable.Prerelease || stable.Build != BuildReady || stable.ImageAvailable != AvailPresent {
		t.Errorf("v2.0.0 = %+v, want latest+ready+present", stable)
	}
	rc := byTag["v2.1.0-rc.1"]
	if !rc.Prerelease || rc.Build != BuildNone || rc.ImageAvailable != AvailUnknown {
		t.Errorf("v2.1.0-rc.1 = %+v, want prerelease+none+unknown", rc)
	}
}

func TestAnnotateReleases_DefaultsUnknownAndNameFallback(t *testing.T) {
	out := AnnotateReleases([]RawRelease{{TagName: "v1.0.0"}}, nil, nil)
	if out[0].ImageAvailable != AvailUnknown {
		t.Errorf("imageAvailable = %v, want unknown when no availability map", out[0].ImageAvailable)
	}
	if out[0].Name != "v1.0.0" {
		t.Errorf("name = %q, want fallback to tag", out[0].Name)
	}
}

func TestAnnotateReleases_LiveMailonFixtures(t *testing.T) {
	releases := loadFixture[[]RawRelease](t, "mailon-releases.json")
	runs := loadFixture[[]RawRun](t, "mailon-runs.json")
	out := AnnotateReleases(releases, runs, nil)
	byTag := map[string]ReleaseAnnotation{}
	for _, r := range out {
		byTag[r.Tag] = r
	}
	// From the captured run fixture: v0.6.9 failed, v0.6.5 ready, v0.6.10 none.
	if byTag["v0.6.9"].Build != BuildFailed {
		t.Errorf("v0.6.9 build = %s, want failed", byTag["v0.6.9"].Build)
	}
	if byTag["v0.6.5"].Build != BuildReady {
		t.Errorf("v0.6.5 build = %s, want ready", byTag["v0.6.5"].Build)
	}
	if byTag["v0.6.10"].Build != BuildNone {
		t.Errorf("v0.6.10 build = %s, want none", byTag["v0.6.10"].Build)
	}
	if !byTag["v0.6.10"].Latest {
		t.Error("v0.6.10 should be the latest published release")
	}
	for _, r := range out {
		if r.ImageAvailable != AvailUnknown {
			t.Errorf("%s availability = %v, want unknown (none supplied)", r.Tag, r.ImageAvailable)
		}
	}
}

// TestReduceRun covers the (status, conclusion) → BuildStatus reduction shared by
// AnnotateBuild and the single-run RunStatus poll (the deploy "wait for build").
func TestReduceRun(t *testing.T) {
	cases := []struct {
		status, conclusion string
		want               BuildStatus
	}{
		{"queued", "", BuildBuilding},
		{"in_progress", "", BuildBuilding},
		{"waiting", "", BuildBuilding},
		{"completed", "success", BuildReady},
		{"completed", "failure", BuildFailed},
		{"completed", "cancelled", BuildFailed},
		{"completed", "timed_out", BuildFailed},
		{"completed", "", BuildFailed},
	}
	for _, c := range cases {
		if got := reduceRun(c.status, c.conclusion); got != c.want {
			t.Errorf("reduceRun(%q,%q) = %v, want %v", c.status, c.conclusion, got, c.want)
		}
	}
}
